### PR TITLE
[TG-2842] Specialization for mocked and unsupported generics

### PR DESCRIPTION
**Pull request by majakusber**
Mocked generic classes and generic classes with unsuppported signatures (e.g., wild cards) are not marked as generic in symbol table. Thus when we see a pointer to them or use them as superclasses/implemented interfaces with some concrete types, we want to fail gracefully by ignoring these concrete types.

### CODE REVIEW

**Comment by majakusber** [comment]
Test-generator PR with regression tests is here: https://github.com/diffblue/test-gen/pull/1613

File: src/java_bytecode/generic_parameter_specialization_map_keys.cpp [hunk 1]
```diff
@@ -99,15 +99,17 @@ const void generic_parameter_specialization_map_keyst::insert_pairs_for_pointer(
       pointer_type.subtype().get(ID_identifier) ==
       pointer_subtype_struct.get(ID_name));
 
-    const java_generic_typet &generic_pointer =
-      to_java_generic_type(pointer_type);
-
     // If the pointer points to an incomplete class, don't treat the generics
-    if(!pointer_subtype_struct.get_bool(ID_incomplete_class))
+    // TODO TG-1996 should treat generic incomplete (mocked) classes
```
Comment (thk123) [inline src/java_bytecode/generic_parameter_specialization_map_keys.cpp]: This comment is quite hard to follow perhaps draw a distinction between stuff that doesn't yet work and stuff that should never work

Comment (thk123) [reply src/java_bytecode/generic_parameter_specialization_map_keys.cpp]: I think possibly the problem is the meaning of "treat the generics" is not clear.

File: unit/goto-programs/goto_program_generics/generic_bases_test.cpp [hunk 2]
```diff
@@ -412,3 +413,99 @@ SCENARIO(
...
+  {
+    const symbol_tablet &symbol_table = load_java_class(
+      "SuperclassUnsupported",
+      "./goto-programs/goto_program_generics",
+      "SuperclassUnsupported.foo");
+
+    THEN("The struct for UnsupportedWrapper1 is complete and non-generic")
+    {
+      const std::string superclass_name = "java::UnsupportedWrapper1";
+      REQUIRE(symbol_table.has_symbol(superclass_name));
```
Comment (thk123) [inline unit/goto-programs/goto_program_generics/generic_bases_test.cpp]: You can use:
```C++
const symbolt &superclass_symbol = 
  require_symbol::require_has_symbol("java::UnsupportedWrapper1", symbol_table);
```
To replace the require and lookup

File: unit/goto-programs/goto_program_generics/GenericHelper.java [hunk 3]
```diff
@@ -23,3 +23,24 @@ public IWrapper(int ii) {
...
+}
+
+// generic class with unsupported signature - multiple bounds
+class UnsupportedWrapper2<T extends InterfaceWrapper & InterfacePairWrapper>
+{
+  public T field;
+}
+
+// generic mocked class, make sure the .class file is not available
+class MockedWrapper<T> {
```
Comment (thk123) [inline unit/goto-programs/goto_program_generics/GenericHelper.java]: I think more consistent to call this `OpaqueWrapper` (particularly in the cbmc context since mocking is a test generation thing)

File: unit/goto-programs/goto_program_generics/generic_bases_test.cpp [hunk 4]
```diff
@@ -412,3 +413,99 @@ SCENARIO(
...
+      "SuperclassMocked.foo");
+
+    THEN("The struct for MockedWrapper is incomplete and not-generic")
+    {
+      const std::string superclass_name = "java::MockedWrapper";
+      REQUIRE(symbol_table.has_symbol(superclass_name));
+
+      const symbolt &superclass_symbol =
+        symbol_table.lookup_ref(superclass_name);
+      const java_class_typet &superclass_type =
```
Comment (thk123) [inline unit/goto-programs/goto_program_generics/generic_bases_test.cpp]: Probably worth pulling this into a `require_type::require_java_class_type` so get a nice error rather than unexpected exception if this fails.

File: unit/goto-programs/goto_program_generics/generic_bases_test.cpp [hunk 5]
```diff
@@ -412,3 +413,99 @@ SCENARIO(
...
+        to_java_class_type(to_class_type(superclass_symbol.type));
+      REQUIRE(
+        to_class_type(superclass_symbol.type).get_bool(ID_incomplete_class));
```
Comment (thk123) [inline unit/goto-programs/goto_program_generics/generic_bases_test.cpp]: `to_class_type(superclass_symbol.type)` -> `superclass_type`?

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 6]
```diff
@@ -202,10 +202,12 @@ void java_bytecode_convert_classt::convert(const classt &c)
       }
       class_type=generic_class_type;
     }
-    catch(unsupported_java_class_signature_exceptiont)
+    catch(unsupported_java_class_signature_exceptiont &e)
```
Comment (peterschrammel) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: `const &` (several occurrences)

File: src/java_bytecode/generic_parameter_specialization_map_keys.cpp (hunk 1 above)
Comment (peterschrammel) [reply src/java_bytecode/generic_parameter_specialization_map_keys.cpp]: Not sure `TG-1996` should be in the code; better move it into the commit message body.

Comment (majakusber) [reply src/java_bytecode/generic_parameter_specialization_map_keys.cpp]: We often include the ticket number with TODOs (especially with disabled or incomplete tests) as searching for a ticket number in the code base seems easier than in the commit messages. As part of making sure that the code is updated when the ticket is done we also add a comment in the ticket (although that not always seems to be true). During our last sprint retrospective we discussed that actually searching for the ticket number in the code base is a good practice to make sure all the tests/code related to a bug are updated and the link to the ticket is removed.

File: src/java_bytecode/generic_parameter_specialization_map_keys.cpp [hunk 7]
```diff
@@ -135,14 +137,19 @@ const void generic_parameter_specialization_map_keyst::insert_pairs_for_symbol(
   const symbol_typet &symbol_type,
   const typet &symbol_struct)
 {
-  if(is_java_generic_symbol_type(symbol_type))
+  // If the class is an incomplete class, don't treat the generics
+  // TODO TG-1996 should treat generic incomplete (mocked) classes
+  // Also do not treat generics is the class is not marked generic or
```
Comment (chrisr-diffblue) [inline src/java_bytecode/generic_parameter_specialization_map_keys.cpp]: 'Also do not treat generics is the class' - 'is' should be 'if' ?

File: unit/goto-programs/goto_program_generics/generic_parameters_test.cpp [hunk 8]
```diff
@@ -486,3 +487,124 @@ SCENARIO(
...
+
+      const symbolt &superclass_symbol =
+        symbol_table.lookup_ref(superclass_name);
+      require_type::require_java_non_generic_class(superclass_symbol.type);
+    }
+
+    WHEN("The method input argument is created in the entry point function")
+    {
+      // For an explanation of this part, look at the comments for the similar
+      // parts of the previous tests.
```
Comment (chrisr-diffblue) [inline unit/goto-programs/goto_program_generics/generic_parameters_test.cpp]: While I realise that this has become something of a boiler plate comment, copy and pasted from other test cases, I actually think there would be more value in copy and pasting the *actual* comment, rather than copy and pasting the reference back to previous tests...

**Review by chrisr-diffblue — COMMENTED** [review]
A few minor comments, and a +1 vote to the comments from @thk123 @peterschrammel

**Comment by majakusber** [comment]
@thk123 @peterschrammel @chrisr-diffblue Thank you for reviews, all should be addressed now. I had to rebase due to failing tests that Chris S. has now fixed but the new changes are in separate commits (starting from 5th commit).

**Review by chrisr-diffblue — APPROVED** [review]
Nothing blocking from my point of view.

**Review by thk123 — APPROVED** [review]
